src/coreclr/tools/aot/ILCompiler.Compiler/IL/TypeEqualityPatternAnalyzer.cs

```diff
@@ -0,0 +1,192 @@
+
+        public readonly bool IsDefault => _state == default;
+        public readonly bool IsTypeEqualityCheck => _state is State.TypeEqualityCheck;
+        public readonly bool IsTypeEqualityBranch => _state is State.Branch;
+        public readonly bool IsTwoTokens => (_flags & Flags.TwoTokens) != 0;
+        public readonly bool IsInequality => (_flags & Flags.Inequality) != 0;
+
+        public void Advance(ILOpcode opcode, in ILReader reader, MethodIL methodIL)
+        {
+            switch (_state)
+            {
+                case 0:
+                    if (opcode == ILOpcode.ldtoken)
+                        (_state, _token1) = (State.LdToken, reader.PeekILToken());
+                    return;
+                case State.LdToken:
+                    if (IsTypeGetTypeFromHandle(opcode, reader, methodIL))
+                        _state = State.TypeOf;
+                    else
+                        break;
+                    return;
+                case State.TypeOf:
+                    if (opcode == ILOpcode.ldtoken)
+                        (_state, _token2) = (State.TypeOf_LdToken, reader.PeekILToken());
+                    else if (IsArgumentOrLocalLoad(opcode))
+                        _state = State.TypeOf_PushedOne;
+                    else if (IsTypeEquals(opcode, reader, methodIL))
+                        _state = State.TypeEqualityCheck;
+                    else if (IsTypeInequals(opcode, reader, methodIL))
+                        (_state, _flags) = (State.TypeEqualityCheck, _flags | Flags.Inequality);
+                    else
+                        break;
+                    return;
+                case State.TypeOf_LdToken:
+                    if (IsTypeGetTypeFromHandle(opcode, reader, methodIL))
+                        _state = State.TypeOf_TypeOf;
+                    else
+                        break;
+                    return;
+                case State.TypeOf_PushedOne:
+                    if (IsObjectGetType(opcode, reader, methodIL))
+                    {
+                        // Nothing, state stays the same
+                    }
+                    else if (IsTypeEquals(opcode, reader, methodIL))
+                        _state = State.TypeEqualityCheck;
+                    else if (IsTypeInequals(opcode, reader, methodIL))
+                        (_state, _flags) = (State.TypeEqualityCheck, _flags | Flags.Inequality);
+                    else
+                        break;
+                    return;
+                case State.TypeOf_TypeOf:
+                    if (IsTypeEquals(opcode, reader, methodIL))
+                        (_state, _flags) = (State.TypeEqualityCheck, _flags | Flags.TwoTokens);
+                    else if (IsTypeInequals(opcode, reader, methodIL))
+                        (_state, _flags) = (State.TypeEqualityCheck, _flags | Flags.TwoTokens | Flags.Inequality);
+                    else
+                        goto case State.TypeOf;
```

Comment:
Do we need to shift the tokens here: `_token1 = _token2;` ?

Reply:
Ah, yes we do.

I was going back and forth between this state machine approach and just emulating stack. Emulating stack would be less bug prone I guess, but we need to model the stack for everything, including calls and that seemed like more of a hassle.

But writing these state machines by hand is also annoying. I started drawing it on a piece of paper at some point.

Reply:
```suggestion
                    {
                        _token1 = _token2;
                        goto case State.TypeOf;
                    }
```